File: csharp/Game/Player.cs
using Godot;
using NodeGetterGenerators;

namespace tutorial;

[VerifyNodeGetters("Player")]
public partial class Player : Area2D
{
    [Node("CollisionShape2D")]
    private partial CollisionShape2D CollisionShape2D { get; }

    [Node("AnimatedSprite2D")]
    private partial PlayerAnimatedSprite2d PlayerAnimatedSprite { get; }

    [Node("GPUParticles2D")]
    private partial GpuParticles2D GpuParticles2D { get; }

    [Export]
    public int Speed { get; set; } = 400;

    [Signal]
    public delegate void HitEventHandler();

    private Vector2 _screenSize;

    public override void _Ready()
    {
        base._Ready();
        _screenSize = GetViewportRect().Size;
        Hide();
        BodyEntered += OnBodyEntered;
    }

    public void Start(Vector2 position)
    {
        Position = position;
        Show();
        CollisionShape2D.SetDeferred(CollisionShape2D.PropertyName.Disabled, false);
        GpuParticles2D.Restart();
    }

    private static Vector2 GetMovementVector()
    {
        var velocity = Vector2.Zero;
        if (Input.IsActionPressed(Constants.Inputs.MoveRight))
            velocity.X += 1;

        if (Input.IsActionPressed(Constants.Inputs.MoveLeft))
            velocity.X -= 1;

        if (Input.IsActionPressed(Constants.Inputs.MoveUp))
            velocity.Y -= 1;

        if (Input.IsActionPressed(Constants.Inputs.MoveDown))
            velocity.Y += 1;

        return velocity;
    }

    private void SetSpriteVelocity(Vector2 velocity)
    {
        var animatedSprite = PlayerAnimatedSprite;
        animatedSprite.SetVelocity(velocity);
        GpuParticles2D.Emitting = animatedSprite.IsPlaying();
    }

    public override void _Process(double delta)
    {
        base._Process(delta);
        var velocity = GetMovementVector();
        if (velocity.Length() > 0)
        {
            velocity = velocity.Normalized() * Speed;
            Position += velocity * (float)delta;
            Position = new Vector2(
                Mathf.Clamp(Position.X, 0, _screenSize.X),
                Mathf.Clamp(Position.Y, 0, _screenSize.Y)
            );
        }
        SetSpriteVelocity(velocity);
    }

    private void OnBodyEntered(Node2D body)
    {
        Hide();
        EmitSignal(SignalName.Hit);
        CollisionShape2D.SetDeferred(CollisionShape2D.PropertyName.Disabled, true);
    }
}

File: swift/swift/Sources/tutorial/Player.swift
import SwiftGodot

@Godot
class Player: Area2D {
    @Export var speed: Int = 400
    var screenSize: Vector2 = Vector2.zero

    @Signal var hit: SimpleSignal

    @Node("CollisionShape2D") private var collisionShape2D: CollisionShape2D
    @Node("AnimatedSprite2D") private var playerAnimatedSprite: PlayerAnimatedSprite
    @Node("GPUParticles2D") private var gpuParticles2D: GPUParticles2D

    func start(position: Vector2) {
        self.position = position
        self.show()
        // TODO: Property name constant
        self.collisionShape2D.setDeferred(property: "disabled", value: Variant(false))
        self.gpuParticles2D.restart()
    }

    private static func getMovementVector() -> Vector2 {
        var velocity = Vector2.zero
        if Input.isActionPressed(action: StringName(Constants.Inputs.moveRight)) {
            velocity.x += 1
        }
        if Input.isActionPressed(action: StringName(Constants.Inputs.moveLeft)) {
            velocity.x -= 1
        }
        if Input.isActionPressed(action: StringName(Constants.Inputs.moveDown)) {
            velocity.y += 1
        }
        if Input.isActionPressed(action: StringName(Constants.Inputs.moveUp)) {
            velocity.y -= 1
        }
        return velocity
    }

    private func setSpriteVelocity(_ velocity: Vector2) {
        self.playerAnimatedSprite.setVelocity(velocity)
        self.gpuParticles2D.emitting = self.playerAnimatedSprite.isPlaying()
    }

    override func _process(delta: Double) {
        super._process(delta: delta)
        var velocity = Player.getMovementVector()
        if velocity.length() > 0 {
            velocity = velocity.normalized() * self.speed
            self.position += velocity * delta
            self.position = Vector2(
                x: self.position.x.clamped(to: 0...self.screenSize.x),
                y: self.position.y.clamped(to: 0...self.screenSize.y))
        }
        self.setSpriteVelocity(velocity)
    }

    override func _ready() {
        self.screenSize = getViewportRect().size
        self.hide()
        self.bodyEntered.connect { body in
            self.onBodyEntered(body)
        }
    }

    private func onBodyEntered(_ body: Node2D?) {
        self.hide()
        self.hit.emit()
        self.collisionShape2D.setDeferred(property: "disabled", value: Variant(true))
    }
}
